lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Implements the plugin manager class.
+
+This module manages each plugin's lifecycle. It is responsible to install, configure and
+upgrade of each of the plugins.
+
+This class is instantiated at the operator level and is called at every relevant event:
+config-changed, upgrade, s3-credentials-changed, etc.
+"""
+
+import logging
+from typing import Any, Dict, List, Optional
+
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchCmdError
+from charms.opensearch.v0.opensearch_plugins import (
+    OpenSearchPlugin,
+    OpenSearchPluginConfig,
+    OpenSearchPluginError,
+    OpenSearchPluginInstallError,
+    OpenSearchPluginMissingDepsError,
+    OpenSearchPluginRemoveError,
+    PluginState,
+)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "da838485175f47dbbbb83d76c07cab4c"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+ConfigExposedPlugins = {}
+
+
+class OpenSearchPluginManager:
+    """Manages plugins."""
+
+    def __init__(self, charm, plugins_path: str = None):
+        """Creates the plugin manager object based on the charm and home_path.
+
+        Stores the home path and, optionally, plugins path can also be passed if it is
+        not available in {home_path}/plugins.
+        """
+        self._charm = charm
+        self._opensearch = charm.opensearch
+        self._opensearch_config = charm.opensearch_config
+        self._charm_config = self._charm.model.config
+        self._plugins_path = plugins_path
+        self._keystore = charm.opensearch_keystore
+
+    @property
+    def plugins(self) -> List[OpenSearchPlugin]:
+        """Returns List of installed plugins."""
+        return [
+            plugin_data["class"](self._plugins_path, extra_config=self._extra_conf(plugin_data))
```

Comment:
This expression looks quite complex and hard to understand tbh.
Maybe it is better to split this into more steps for the sake of readability?